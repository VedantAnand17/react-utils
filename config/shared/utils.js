/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
