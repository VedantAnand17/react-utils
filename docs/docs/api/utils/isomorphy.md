# isomorphy
```js
import { isomorphy } from '@dr.pogodin/react-utils';
```
The [isomorphy] module is a collection of helpers for dealing with isomorphic
aspects of code (mostly environment checks: client- or server-side, development
or production, _etc._). All [constants](#constants) and [methods](#methods)
documented below are exposed as key values of `isomorphy` object.

:::caution Beware
This module relies on [Babel](/docs/api/configs/babel) and
[Webpack](/docs/api/configs/webpack) configurations provided
by **react-utils** library being used by the host code, along with provided
mechanics for server creation and client-side app initialization. It won't
work correctly with a different app setup.
:::

## Constants
- `isomorphy.IS_CLIENT_SIDE` - **boolean** - equals **true** within the client-side
  (browser) environment, and equals **false** at the server-side (NodeJS).
- `isomorphy.IS_SERVER_SIDE` - **boolean** - equals **true** within the server-side
  (NodeJS) environment, and equals **false** at the client-side (browser).

## Methods
- [buildTimestamp()](#buildtimestamp) - Returns the build timestamp of
  the frontend JS bundle.
- [getBuildInfo()](#getbuildinfo) - Returns the build info object.
- [isDevBuild()](#isdevbuild) - Returns **true** if the development version of
  code is running.
- [isProdBuild()](#isprodbuild) - Returns **true** if the production version of
  code is running.

### buildTimestamp()
```jsx
isomorphy.buildTimestamp() => string
```
Returns **string**: the build timestamp of the frontend JS bundle in ISO format.

### getBuildInfo()
```jsx
isomorphy.getBuildInfo() => object
```
Returns **object**: the build info object.

### isDevBuild()
```jsx
isomorphy.isDevBuild() => boolean
```
Returns **boolean**: **true** if the development version of code is running,
**false** otherwise.

### isProdBuild()
```jsx
isomorphy.isProdBuild() => boolean
```
Returns **boolean**: **true** if the production version of code is running,
**false** otherwise.

[Isomorphy]: /docs/api/utils/isomorphy
