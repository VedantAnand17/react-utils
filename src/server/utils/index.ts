// Aggregates utility modules together.

export * as errors from './errors';
