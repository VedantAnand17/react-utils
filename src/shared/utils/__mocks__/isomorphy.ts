/**
 * Mockup of the isomorphy module.
 */
/* global jest */

const MOCK_BUILD_TIMESTAMP = 'Wed, 29 Nov 2017 07:40:00 GMT';

const mock = jest.requireActual('../isomorphy');

mock.buildTimestamp = () => MOCK_BUILD_TIMESTAMP;

module.exports = mock;
