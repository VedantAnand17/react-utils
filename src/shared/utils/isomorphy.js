/**
 * @category Utilities
 * @module isomorphy
 * @desc
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * ```
 * Collection of helpers for dealing with isomorphic aspects of code
 * (mostly environment checks: client- or server-side, development or
 * production, _etc._).
 *
 * **Beware:** This module relies on Babel and Webpack configurations provided
 * by `react-utils` library being used by the host code, along with provided
 * methods for server creation and client-side app initialization. It won't
 * work correctly with a different app setup.
 */
/* global window */

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { IS_CLIENT_SIDE } = isomorphy;
 * ```
 * `true` within client-side environment (browser), `false` at server-side.
 */
export const IS_CLIENT_SIDE = typeof process !== 'object'
  || !process.versions || !process.versions.node;

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { IS_SERVER_SIDE } = isomorphy;
 * ```
 * `true` within the server-side environment (node), `false` at client-side.
 */
export const IS_SERVER_SIDE = !IS_CLIENT_SIDE;

/**
 * @ignore
 * @return {string} Code mode: "development" or "production".
 */
function getMode() {
  return process.env.NODE_ENV;
}

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { isDevBuild } = isomorphy;
 * ```
 * Returns `true` if development version of the code is running;
 * `false` otherwise.
 * @return {boolean}
 */
export function isDevBuild() {
  return getMode() === 'development';
}

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { isProdBuild } = isomorphy;
 * ```
 * Returns `true` if production build of the code is running;
 * `false` otherwise.
 * @return {boolean}
 */
export function isProdBuild() {
  return getMode() === 'production';
}

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { getBuildInfo } = isomorphy;
 * ```
 * Returns build info object.
 * @returns {object}
 */
export function getBuildInfo() {
  return (IS_CLIENT_SIDE ? window : global).TRU_BUILD_INFO;
}

/**
 * ```js
 * import { isomorphy } from '@dr.pogodin/react-utils';
 * const { buildTimestamp } = isomorphy;
 * ```
 * Returns build timestamp of the front-end JS bundle.
 * @return {string} ISO date/time string.
 */
export function buildTimestamp() {
  return getBuildInfo().timestamp;
}
