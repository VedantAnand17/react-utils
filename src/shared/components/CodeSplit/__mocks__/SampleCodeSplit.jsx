/**
 * This is a sample component, doing a code split.
 */

import React from 'react';
import path from 'path';

import CodeSplit from 'components/CodeSplit';

import { hooks, webpack } from 'utils';

export default function SampleCodeSplit() {
  const [testKey] = hooks.useGlobalState('test.key', 'testValue');
  return (
    <>
      <h1>{testKey}</h1>
      <CodeSplit
        chunkName="sample-component-1"
        getComponentAsync={
          () => import(/* webpackChunkName: 'sample-component-1' */ './SampleComponent')
        }
      />
      <CodeSplit
        chunkName="sample-component-2"
        getComponentAsync={
          () => import(/* webpackChunkName: 'sample-component-2' */ './SampleComponent')
        }
        placeholder={() => <div>PLACEHOLDER!</div>}
      />
      <CodeSplit
        chunkName="sample-component-3"
        getComponentAsync={
          () => import(/* webpackChunkName: 'sample-component-3' */ './SampleComponent')
        }
        getComponentServer={() => {
          const p = webpack.resolveWeak('components/CodeSplit/__mocks__/SampleComponent');
          return webpack.requireWeak(path.resolve(__dirname, p));
        }}
        placeholder={() => <div>PLACEHOLDER!</div>}
      />
    </>
  );
}
