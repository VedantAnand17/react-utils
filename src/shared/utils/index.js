import path from 'path';

import config from './config';
import * as isomorphy from './isomorphy';
import * as redux from './redux';
import * as webpack from './webpack';

const juUrl = module.webpackPolyfill ? './shared/utils/jest'
  : path.resolve(__dirname, './jest');

const JU = isomorphy.isServerSide() ? webpack.requireWeak(juUrl) : null;

export {
  config,
  isomorphy,
  JU,
  redux,
  webpack,
};
