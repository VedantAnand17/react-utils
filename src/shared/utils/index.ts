import themedImpl, {
  type Theme,
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

import type * as JuT from './jest';

export {
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

// Note: it should be done this way, as in some environments
// "process" might not exist, and process.env.NODE_CONFIG_ENV
// not injected by Webpack.
let NODE_CONFIG_ENV;
try {
  NODE_CONFIG_ENV = process.env.NODE_CONFIG_ENV;
} catch { /* noop */ }

const env = NODE_CONFIG_ENV || process.env.NODE_ENV;

const JU: typeof JuT | null = env !== 'production'
  ? webpack.requireWeak('./jest', __dirname) as typeof JuT | null
  : null;

export {
  type Theme,
  config,
  isomorphy,
  JU,
  themed,
  ThemeProvider,
  time,
  webpack,
};
