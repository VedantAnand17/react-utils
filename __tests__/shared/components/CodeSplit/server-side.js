/**
 * Here are the actual code splitting tests.
 */

import mockdate from 'mockdate';
import pretty from 'pretty';

import { IS_SERVER_SIDE } from 'utils/isomorphy';

import SampleCodeSplit from './__assets__/SampleCodeSplit';
import { renderServerSide } from './__assets__/shared';

jest.mock('node-forge');
jest.mock('uuid');

jest.useFakeTimers();
mockdate.set('2019-11-29Z');

test('Server-side rendering', async () => {
  expect(IS_SERVER_SIDE).toBe(true);
  let markup = await renderServerSide(SampleCodeSplit, 1);
  expect(pretty(markup)).toMatchSnapshot();
  markup = renderServerSide(SampleCodeSplit, 3);
  jest.runAllTimers();
  markup = await markup;
  expect(pretty(markup)).toMatchSnapshot();
});
