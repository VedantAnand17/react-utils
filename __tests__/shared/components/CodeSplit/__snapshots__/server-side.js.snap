// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server-side rendering 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

  <head>

    <title data-react-helmet=\\"true\\"></title>

    <meta name=\\"theme-color\\" content=\\"#FFFFFF\\" />

    <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-1-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-2-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-3-1511941200000.css\\" rel=\\"stylesheet\\" />

    <meta charset=\\"utf-8\\" />
    <meta content=\\"width=device-width,initial-scale=1.0\\" name=\\"viewport\\" />
  </head>

  <body>

    <div id=\\"react-view\\">
      <h1>testValue</h1>
      <div data-chunk-name=\\"sample-component-1\\"></div>
      <div data-chunk-name=\\"sample-component-2\\">
        <div>PLACEHOLDER!</div>
      </div>
      <div data-chunk-name=\\"sample-component-3\\">
        <div>
          <h1>Sample Component</h1><pre>{
  &quot;test&quot;: {
    &quot;key&quot;: &quot;testValue&quot;
  },
  &quot;dr_pogodin_react_utils___split_components&quot;: {
    &quot;sample-component-1&quot;: {
      &quot;data&quot;: null,
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;S00000000-0000-0000-0000-000000000001&quot;,
      &quot;timestamp&quot;: 0
    },
    &quot;sample-component-2&quot;: {
      &quot;data&quot;: null,
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;S00000000-0000-0000-0000-000000000002&quot;,
      &quot;timestamp&quot;: 0
    }
  }
}</pre>
        </div>
      </div>
    </div>
    <script id=\\"inj\\" type=\\"application/javascript\\" nonce=\\"abcdef-dummy-nonce\\">
      window.INJ = \\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBb1VDFWdXsViFyeX13tuynQ88EETx6RuQyms6aHDjsZzm54u6pdtE5bElWjYQQYYKWQ6Q7AG6jOJ0m4jyHJEL2hbRtKnBKIGH+RZ8NtSPyZUdQIHOOyAmEjNWwTcoflKwqX1yZQFqY61t71j0sQev32jJQxdFCAfnIPEBbQyoMOthnONT2MDoAhYBo7mW5WvIIwKZOdbtEV00hm8UfXRtHfiDZlOuY93yXGAlvaXY18BqM3DPSi818K7Z1H4SWC1VYTGrNkXoMDIE6ybStZpZjR4spwNSQv32yICqmIFSe5L6VP07D+zuaRT8l8Kisyjw5b2Xh3ZZWsXzZjicu32lyebon9uckaiJy3Jj8fVO83oGnoVoG23K8ksWQe5vTPxUvDNJMayc7CzFtcUZFtdicQT0RPhKYtcvngtzc19MY6rKHS5iMFX5Ms6j+LMWfYk3Np31TfQ3q5XhqvVHZPciRN1TTzEdcfcEJ+d8oQ+6GKM=\\"
    </script>
    <script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
    <script src=\\"/test/public/path/sample-component-2-1511941200000.js\\" type=\\"application/javascript\\"></script>
    <script src=\\"/test/public/path/sample-component-3-1511941200000.js\\" type=\\"application/javascript\\"></script>

  </body>

</html>"
`;

exports[`Server-side rendering 2`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

  <head>

    <title data-react-helmet=\\"true\\"></title>

    <meta name=\\"theme-color\\" content=\\"#FFFFFF\\" />

    <link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-1-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-2-1511941200000.css\\" rel=\\"stylesheet\\" />
    <link href=\\"/test/public/path/sample-component-3-1511941200000.css\\" rel=\\"stylesheet\\" />

    <meta charset=\\"utf-8\\" />
    <meta content=\\"width=device-width,initial-scale=1.0\\" name=\\"viewport\\" />
  </head>

  <body>

    <div id=\\"react-view\\">
      <h1>testValue</h1>
      <div data-chunk-name=\\"sample-component-1\\">
        <div>
          <h1>Sample Component</h1><pre>{
  &quot;test&quot;: {
    &quot;key&quot;: &quot;testValue&quot;
  },
  &quot;dr_pogodin_react_utils___split_components&quot;: {
    &quot;sample-component-1&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    },
    &quot;sample-component-2&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    }
  }
}</pre>
        </div>
      </div>
      <div data-chunk-name=\\"sample-component-2\\">
        <div>
          <h1>Sample Component</h1><pre>{
  &quot;test&quot;: {
    &quot;key&quot;: &quot;testValue&quot;
  },
  &quot;dr_pogodin_react_utils___split_components&quot;: {
    &quot;sample-component-1&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    },
    &quot;sample-component-2&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    }
  }
}</pre>
        </div>
      </div>
      <div data-chunk-name=\\"sample-component-3\\">
        <div>
          <h1>Sample Component</h1><pre>{
  &quot;test&quot;: {
    &quot;key&quot;: &quot;testValue&quot;
  },
  &quot;dr_pogodin_react_utils___split_components&quot;: {
    &quot;sample-component-1&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    },
    &quot;sample-component-2&quot;: {
      &quot;data&quot;: {},
      &quot;numRefs&quot;: 0,
      &quot;operationId&quot;: &quot;&quot;,
      &quot;timestamp&quot;: 1574985600000
    }
  }
}</pre>
        </div>
      </div>
    </div>
    <script id=\\"inj\\" type=\\"application/javascript\\" nonce=\\"abcdef-dummy-nonce\\">
      window.INJ = \\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBb1VDFWdXsViFyeX13tuynQ88EETx6RuQyms6aHDjsZzm54u6pdtE5bElWjYQQYYKWQ6Q7AG6jOJ0m4jyHJEL2hbRtKnBKIGH+RZ8NtSPyZUdQIHOOyAmEjNWwTcoflKwqX1yZQFqY61t71j0sQev32jJQxdFCAfnIPEBbQyoMOthnONT2MDoAhYBo7mW5WvIIwKZOdbtEV00hm8UfXRtHfiDZlOuY93yXGAlvaXY18BqM3DPSi818K7Z1H4SWC1VYTGrNkXoMDIE6ybStZpZjR4spwNSQv32yICqmIFSe5L6VP07D+zuaRT8l8Kisyjw5b2Xh3ZZWsXzZjicu32lyebon9uckaiJy3Jj8fVO83oGnoVoG23K8ksWQe5vTPxUvDNJMayc7CzFtcUZFtdicQT0RPhKYtcvngtzc19MY6rKHS5iMFX5Ms6j+LMWfYk3Np31TfQ3q5XhqvVHZPciRN1TTzEdcfcEJ+d8oQ+6GKM=\\"
    </script>
    <script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
    <script src=\\"/test/public/path/sample-component-2-1511941200000.js\\" type=\\"application/javascript\\"></script>
    <script src=\\"/test/public/path/sample-component-3-1511941200000.js\\" type=\\"application/javascript\\"></script>

  </body>

</html>"
`;
